Added clearing of cache groups

pbearne opened the pull request:
https://github.com/WordPress/performance/issues/38
https://core.trac.wordpress.org/ticket/4476
https://core.trac.wordpress.org/ticket/9773
solves #INITS-115



Trac ticket: https://core.trac.wordpress.org/ticket/4476
https://core.trac.wordpress.org/ticket/9773)

pbearne committed:
Added clearing of cache groups
https://github.com/WordPress/performance/issues/38
https://core.trac.wordpress.org/ticket/4476
https://core.trac.wordpress.org/ticket/9773
solves #INITS-115

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,45 @@ function wp_cache_flush() {
+		foreach ( $group as $group_key ) {
+
+			wp_cache_flush_group( $group_key );
+		}
```
```suggestion
		array_map( 'wp_cache_flush_group', $group );
```

Could we use array map here?

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,45 @@ function wp_cache_flush() {
+		}
+
+		return $wp_object_cache->flush_group( $group );
+	}
```
I am not sure I understand this. Why also check these link groups?

If we are going to do this, then, all meta group would need to be cleared. So post, comment, site and term. Not to mention other linked groups, like network and site (network) options. If we do this, we would need a way to register the links together, so that developers could link custom groups together.

pbearne replied:
I found a comment about this linked group and felt it was safer for the core to handle linked groups rather than rely on coders knowing (I didn't) about the links
So yes I believe we should handle all the linked groups I just don't know them all

pbearne replied:
and I should change the code to use an array to hold the mappings

pbearne replied:
I have refactored this to allow to adding additional mappings 
@spacedmonkey can you help with the pairs we need to flush together

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,45 @@ function wp_cache_flush() {
+		return $wp_object_cache->flush_group( $group );
+	}
+
+	return $wp_object_cache->flush();
```
So if flush group doesn't exist, we just flush everything?

pbearne replied:
yes I felt that should be the correct action
assume that if tried to flush a group is was because you needed to clear it and were be light in your action
If we don't do we will need to return false and suggest the calling code handles it

we could provide an option "flush_all_if_no_group_flush"!

pbearne replied:
removed

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

sboisvert commented:
@pbearne I worry about the flush cache group automatically flushing everything. I like the approach here: https://github.com/pantheon-systems/wp-redis/commit/db32e946ed62b5736ef34a938f66966e1fcf6423 that only flushes if it's possible to target them. It feels like a safer approach.

What we could do is, separately from the changes to core. Add support for groups to object cache libraries. The redis one above already implements it but I don't think any memcache ones do. I suspect it's because memcache has no concept of groups and therefore it's just a hash of the concatenation of the key and group most of the time, but we could pseudo implement it by having groups be a pointer to an incrementor and when you flush the cache group you're in essence just incrementing the group part of the cache key, which while not actually flushing the keys, would make them return false / repopulate them.

pbearne committed:
removed full cache if group cache fails
added filter to allow plugin to adjust the called function

pbearne committed:
whitespace

pbearne committed:
comments

pbearne commented:
@sboisvert I have removed the failback
I have refactored the functions to make them more SOLID 
I will need to add some more tests for the new functions but I through I get the code/approach sorted first
Should a change the method to detect and function names to redist example?

pbearne committed:
refactor function and new test for new functions

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,110 @@ function wp_cache_flush() {
+		foreach ( $cache_pairs[ $group ] as $cache_pair ) {
+			wp_cache_flush_group( $cache_pair, false );
+		}
+	}
```
I strongly believe that all this linked group logic should be removed. I don't believe it makes sense. If you want to clear a group, you should just call `wp_cache_flush_group` multiple times. Similar to [this](https://github.com/WordPress/wordpress-develop/blob/ea1fe31f028156a56eb97e8b7cce3d0f0e6e242f/src/wp-includes/post.php#L7400-L7401). 

Flushing this linked groups may have side effects we do not wish for. Let's remove this for now. Maybe we could add it back later.

pbearne replied:
removed

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,110 @@ function wp_cache_flush() {
+	 *
+	 * @param string return $flush_group_function name of function to call from the cashing class.
+	 */
+	$flush_group_function = apply_filters( 'wp_cache_flush_group_function', 'flush_group' );
```
Not a function, but a method right?

spacedmonkey replied:
What is the benefit of this filter?

pbearne replied:
this allowed plugins to change function that handles the flushing 
removed

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,110 @@ function wp_cache_flush() {
+	 * @param string return $flush_group_function name of function to call from the cashing class.
+	 */
+	$flush_group_function = apply_filters( 'wp_cache_flush_group_function', 'flush_group' );
+	return $wp_object_cache->$flush_group_function( $group );
```
Need to check if method is callbable.

pbearne replied:
added

pbearne committed:
dumped the function down

pbearne committed:
removed not needed tests

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,62 @@ function wp_cache_flush() {
+			'user_meta' => array( 'users' ),
+		)
+	);
+}
```
This function is not needed for this patch right?

pbearne replied:
yes I will remove it

pbearne committed:
removed not needed function

tillkruss commented on src/wp-includes/cache.php:
```diff
@@ -102,6 +102,35 @@ function wp_cache_flush() {
+	}
+
+	return false;
+}
```
This needs a `cache-compat.php` equivalent, no?

pbearne replied:
I am not if we do it in cache-compat.php as we need support in the caching tool to support this

pbearne replied:
added

pbearne committed:
Merge branch 'trunk' into INITS-115-Support_flushing_cache_groups

# Conflicts:
#	src/wp-includes/class-wp-object-cache.php
#	tests/phpunit/tests/cache.php

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/class-wp-object-cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/class-wp-object-cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Added fallback to cache-compat.php

pbearne committed:
Merge remote-tracking branch 'origin/INITS-115-Support_flushing_cache_groups' into INITS-115-Support_flushing_cache_groups

pbearne committed:
Added wp_cache_flush_group to get_pluggable_function_signatures test

tillkruss commented on src/wp-includes/cache.php:
```diff
@@ -131,6 +131,34 @@ function wp_cache_set_multiple( array $data, $group = '', $expire = 0 ) {
+	if ( is_array( $group ) ) {
+		array_map( 'wp_cache_flush_group', array_values( $group ) );
+
+		return true;
```
This is a bit ambiguous. Should we return an array of booleans, if an array is given? @spacedmonkey?

pbearne committed:
Added array return if array is passed

pbearne committed:
removed not needed test wp_cache_get_linked_meta

pbearne committed:
removed whitespace changes

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache-compat.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Merge branch 'WordPress:trunk' into INITS-115-Support_flushing_cache_groups

pbearne committed:
Update tests/phpunit/tests/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

spacedmonkey commented on src/wp-includes/cache-compat.php:
```diff
@@ -141,3 +141,21 @@ function wp_cache_flush_runtime() {
+	 * @return bool|Array True or array of bool in array passed on success, false on failure group not found.
+	 */
+	function wp_cache_flush_group( $group ) {
+		return wp_cache_flush();
```
Maybe this should have a doing it wrong or something here. Tell developers to no call this function unless your objet cache supports it.

tillkruss replied:
Yeah, good call!

pbearne replied:
added

pbearne committed:
Update tests/phpunit/tests/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

spacedmonkey commented on src/wp-includes/cache-compat.php:
```diff
@@ -141,3 +141,21 @@ function wp_cache_flush_runtime() {
+	 */
+	function wp_cache_flush_group( $group ) {
+		return wp_cache_flush();
+	}
```
```suggestion
	function wp_cache_flush_group( $group ) {
	        global $wp_object_cache;
	        
	        if( $wp_object_cache && method_exists( $wp_object_cache, 'flush_group' ) ) {
	            return $wp_object_cache->flush_group( $group );
	        }
	        
	        _doing_it_wrong( __FUNCTION__, __( 'Your implementation of object caching does not support flush by group.' ), '6.0.0' );
	        
		return wp_cache_flush();
	}
```

spacedmonkey commented on src/wp-includes/cache.php:
```diff
@@ -132,6 +132,29 @@ function wp_cache_set_multiple( array $data, $group = '', $expire = 0 ) {
+		return array_map( 'wp_cache_flush_group', array_values( $group ) );
+	}
+
+	return $wp_object_cache->flush_group( $group );
```
Do we need to ensure that this method exists or will it always exist?

tillkruss replied:
It will always exist, because it links to `class-wp-object-cache.php`. Plugins will get the compat fallback method.

spacedmonkey commented on tests/phpunit/tests/cache.php:
```diff
@@ -378,7 +378,7 @@ public function test_wp_cache_set_multiple() {
 	 * @ticket 20875
 	 */
-	public function test_wp_cache_get_multiple() {
+	public function test_get_multiple() {
```
Let's not change unrelated test names in this PR.

pbearne replied:
reverted

spacedmonkey commented on tests/phpunit/tests/cache.php:
```diff
@@ -415,4 +415,71 @@ public function test_wp_cache_delete_multiple() {
+	}
+	public function set() {
+		return;
+	}
```
Let's move this class to it's own file that lives in `tests/phpunit/includes`

spacedmonkey commented on tests/phpunit/tests/cache.php:
```diff
@@ -415,4 +415,71 @@ public function test_wp_cache_delete_multiple() {
+	public function test_wp_cache_flush_group() {
+		if ( ! method_exists( $this->cache, 'flush_group' ) ) {
+			$this->markTestSkipped( 'This test requires a flush_group() method on the cache object.' );
+		}
```
This function should always work, so we don't need to skip it.

tillkruss replied:
Then the `tests/phpunit/includes/object-cache.php` needs to implement it.

spacedmonkey replied:
What is the point of the test then?

spacedmonkey commented on tests/phpunit/tests/cache.php:
```diff
@@ -415,4 +415,71 @@ public function test_wp_cache_delete_multiple() {
+		$this->assertSame( $val, wp_cache_get( $key, 'group-kept' ), 'test_wp_cache_flush_groups: group-kept should still contain my-val' );
+	}
+}
+class WP_Object_Cache_dummy {
```
What does this do again?

pbearne committed:
Update tests/phpunit/tests/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache-compat.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
moved helper class to includes
reverted test rename

pbearne committed:
Merge branch 'WordPress:trunk' into INITS-115-Support_flushing_cache_groups

spacedmonkey commented:
You might want to revert these commits and just merge trunk into this branch. I am seeing a lot of changes that are unrelated to this PR.

pbearne committed:
removed method exists check
removed WP_Object_Cache_dummy class

pbearne committed:
removed method exists check
removed WP_Object_Cache_dummy class

spacedmonkey reviewed (approved):
This patch looks solid now. Let's submit it to WP Core team and get feedback from the wider community.

pbearne committed:
Merge branch 'WordPress:trunk' into INITS-115-Support_flushing_cache_groups

pbearne committed:
whitespaces

pbearne committed:
added flush_group to memcache test code

pbearne committed:
added flush_group to memcache test code

pbearne committed:
added check for method before calling

pbearne committed:
added check for method before calling
And flush all if flush_group not supported

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache-compat.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache-compat.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/class-wp-object-cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
added constant WP_OBJECT_CACHE_SUPPORTS_GROUP_FLUSH

pbearne committed:
Update src/wp-includes/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/cache-compat.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Added WP_OBJECT_CACHE_SUPPORTS_GROUP_FLUSH to the memcache test cache file and set to false

pbearne committed:
fixed whitespace

pbearne committed:
Update src/wp-includes/cache-compat.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/cache.php

Co-authored-by: Till Krüss <tillkruss@users.noreply.github.com>
